src/SFA.DAS.EmployerIncentives.Commands/Persistence/ApprenticeshipIncentiveDomainRepository.cs

```diff
@@ -0,0 +1,53 @@
+﻿using SFA.DAS.EmployerIncentives.Abstractions.Events;
+using SFA.DAS.EmployerIncentives.Data.ApprenticeshipIncentives;
+using SFA.DAS.EmployerIncentives.Domain.Factories;
+using System;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EmployerIncentives.Commands.Persistence
+{
+    public class ApprenticeshipIncentiveDomainRepository : IApprenticeshipIncentiveDomainRepository
+    {
+        private readonly IApprenticeshipIncentiveDataRepository _apprenticeshipIncentiveDataRepository;
+        private readonly IApprenticeshipIncentiveFactory _apprenticeshipIncentiveFactory;
+        private readonly IDomainEventDispatcher _domainEventDispatcher;
+
+        public ApprenticeshipIncentiveDomainRepository(
+            IApprenticeshipIncentiveDataRepository apprenticeshipIncentiveDataRepository,
+            IApprenticeshipIncentiveFactory apprenticeshipIncentiveFactory,
+            IDomainEventDispatcher domainEventDispatcher)
+        {
+            _apprenticeshipIncentiveDataRepository = apprenticeshipIncentiveDataRepository;
+            _apprenticeshipIncentiveFactory = apprenticeshipIncentiveFactory;
+            _domainEventDispatcher = domainEventDispatcher;
+        }
+
+        public async Task<Domain.ApprenticeshipIncentives.ApprenticeshipIncentive> Find(Guid id)
+        {
+            var application = await _apprenticeshipIncentiveDataRepository.Get(id);
+            if (application != null)
+            {
+                return await Task.FromResult(_apprenticeshipIncentiveFactory.GetExisting(id, application));
+            }
+
+            return null;
+        }
+
+        public async Task Save(Domain.ApprenticeshipIncentives.ApprenticeshipIncentive aggregate)
+        {
+            if (aggregate.IsNew)
+            {
+                await _apprenticeshipIncentiveDataRepository.Add(aggregate.GetModel());
+            }
+            else
+            {
+                await _apprenticeshipIncentiveDataRepository.Update(aggregate.GetModel());
```

Comment:
Should this set IsNew to false to it doesn't call add again if Save is called twice?

Reply:
The distributed lock should prevent this from being called twice.  We don't persist the isNew state of the domain object, it's isNew true when created else false if the state is loaded from the db.  It's a good point but if called twice they would have  two different Id's anyway.